### ISSUE TITLE
no-shorthand + ansible.builtin.raw: The raw module takes a free form command to run.

### ISSUE THREAD

**Issue by hswong3i** [ansible/ansible-lint#2537]
##### Summary

https://github.com/ansible/ansible-lint/pull/2527 introduce ```no-shorthand``` rule as experimental, but https://docs.ansible.com/ansible/latest/collections/ansible/builtin/raw_module.html only support free form arguments.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
ansible --version
ansible [core 2.13.4]
  config file = None
  configured module search path = ['/home/hswong3i/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/hswong3i/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.6 (main, Aug 10 2022, 11:40:04) [GCC 11.3.0]
  jinja version = 3.1.2
  libyaml = True

ansible-lint --version
ansible-lint 6.8.0b1 using ansible 2.13.4
```

- ansible installation method: OS package (https://build.opensuse.org/package/show/home:alvistack/ansible-community-ansible-build-data-6.4.0)
- ansible-lint installation method: OS package (https://build.opensuse.org/package/show/home:alvistack/ansible-community-ansible-lint-6.8.0-beta1)

##### OS / ENVIRONMENT

```
$ cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

##### STEPS TO REPRODUCE

Consider follow task with raw module for bootstrapping python3 (https://github.com/alvistack/vagrant-ubuntu/blob/597240a8b75db0d286f58861e398f104d431143b/playbooks/prepare.yml#L35-L43):
```
    - name: debian | apt-get install python3
      ansible.builtin.raw: |
        set -eu
        apt-get update
        DEBIAN_FRONTEND=noninteractive apt-get install -y python3
      args:
        executable: "/bin/bash"
      changed_when: false
      failed_when: false
```

ansible-lint 6.8.0b1 is now complain as below:
```
no-shorthand: Avoid using shorthand (free-form) when calling module actions. (ansible.builtin.raw) (no-shorthand)
vagrant-ubuntu/playbooks/prepare.yml:35 Task/Handler: debian | apt-get install python3
```

**Comment by konstruktoid** [comment]
Same situation with `shell`, https://github.com/konstruktoid/ansible-role-hardening/actions/runs/3169895629/jobs/5162110233#step:4:3941:
```console
no-shorthand: Avoid using shorthand (free-form) when calling module actions. (ansible.builtin.shell) (no-shorthand)
molecule/default/verify.yml:460 Task/Handler: Verify RedHat GRUB IPv6 settings
```

https://github.com/konstruktoid/ansible-role-hardening/blob/master/molecule/default/verify.yml#L460-L468

**Comment by hswong3i** [comment]
@konstruktoid case different for shell, because it have cmd for long hand, see https://docs.ansible.com/ansible/latest/collections/ansible/builtin/shell_module.html#parameter-cmd

This is also the original though for this short hand rule, else it will become a chaos for lint.

But here the case for raw is different, it ONLY support free form so we have no choice but asking ansible-lint to exclude it.

**Comment by konstruktoid** [comment]
> @konstruktoid case different for shell, because it have cmd for long hand, see https://docs.ansible.com/ansible/latest/collections/ansible/builtin/shell_module.html#parameter-cmd
> 

Absolutely correct, my bad.

**Comment by ssbarnea** [comment]
Clearly that we will add exceptions ASAP if there is no correct way to avoid it.

`raw` is interesting because it should be possible to write `raw: executable=/bin/bash echo $SHELL`, which look like a valid case for triggering the rule.

At the same time, I do see the example you give as very good reason for doing something to avoid that false positive. I guess that looking for an `=` sign was not good enough...

**Comment by ssbarnea** [comment]
Apparently it is impossible to call `raw` without engaging the "shorthand" behavior in Ansible, which IMHO makes is quite prone to errors, as any presence on `executable=...` inside the passed string will likely mess the behavior.

To make it even more interesting I discovered that ansible accepts anything you pass to raw, even an entire dictionary. What happens is that it seems to pass without any error.

```yaml
    - name: Invalid task which ansible accepts (returns nothing)
      ansible.builtin.raw:
        foo: 1
        br: xxxxx
      args:
        executable: /bin/bash
```

Ansible reports success on this:
```
changed: [localhost] => 
    changed: true
    rc: 0
    stderr: ''
    stdout: ''
```

I guess that for the moment the only thing I can do is to add a special check for `raw` which will trigger the issue only if `executable=` is inside the string.

**Comment by hswong3i** [comment]
> I guess that for the moment the only thing I can do is to add a special check for raw which will trigger the issue only if executable= is inside the string.

It looks like a good idea:
1. Messing executable inside free form for raw is crazy…
2. Official example passing the only available argument executable with args, similar as my above code snippet too
3. Passing executable with args also valid across both shell command or raw so we could keep code pattern as similar as possible
4. Regex for checking executable= also looks clean and elegant ;-)

**Comment by bcoca** [comment]
Normally 'shorthand' refers to YAML shorthand  `[1,2,3]` vs `- 1\n- 2\n\- 3`. k=v has been called 'inline' or 'key value pairs' or 'equals separated/notation'

**Comment by hswong3i** [comment]
@ssbarnea I give a quick fixup as https://build.opensuse.org/package/show/home:alvistack/ansible-community-ansible-lint-6.8.0-beta1+2022100332598c51 and now no more false negative for raw:
```
$ ansible-lint --version
ansible-lint 6.8.0b1+2022100332598c51 using ansible 2.13.4
```

**Comment by bcoca** [comment]
also note that 'inline paramters' can only be passed via `varname=value`, so this can break includes/imports/roles

**Comment by ssbarnea** [comment]
@bcoca Can you please elaborate on `inline parameters` part? I was not able to find any example using that on official documentation:

* http://docs.ansible.com/ansible/latest/collections/ansible/builtin/include_role_module.html#ansible-collections-ansible-builtin-include-role-module
* http://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html

**Comment by bcoca** [comment]
it is the 2nd highest precedence, specially for roles aka 'include params' but we normally call them 'inline' cause k=v format

**Comment by ssbarnea** [comment]
The real problem here is lack of documentation. Apparently that key feature is missing from the main pages of docs and on the [variable precedence page](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#ansible-variable-precedence) is only briefly mentioned, without even explaining what it means.

There are the only places where `params` can be found on the entire page, clearly not an indication of some
![](https://sbarnea.com/ss/Screen-Shot-2022-10-03-17-56-44.10.png)

I am not very keen to implement workarounds for behaviors that are mostly undocumented.

TBH, I was not able to call any role using free-form, at least with ansible 2.14.0b1, free form did not work for `role:`, neither for `include_role`. So, I think that the only exception is for now with `raw:`.

**Comment by bcoca** [comment]
they WERE documented, but for the sake of 'normalization' they were converted ... same thing that is happening TODAY with booleans being normalized into True/False

**Comment by apatard** [comment]
> they WERE documented, but for the sake of 'normalization' they were converted ... 

But I thought that at least include params were not supported anymore ? eg:

```
  - include_tasks: tasks.yml myparam="value"
```

> same thing that is happening TODAY with booleans being normalized into True/False

Well, for booleans, it's a different story imho. When I discovered about true/false only instead of yes/no/true/false, I got annoyed (bunch of places to fix) but it's YAML 1.2 forcing that on us, so complain to YAML devs. While some parsers are still ok with YAML 1.1, it will possibly not always be true, so it's not a bad idea to start fixing that now...

**Comment by bcoca** [comment]
yes, for includes it got removed, but should still work for roles

on the true/false (just an example, many others of normalizing docs making 'features disappear') YAML .2 is not what Ansible uses, pyyaml is still YAML 1.1
